include/seqan3/alignment/pairwise/policy/alignment_matrix_policy.hpp

```diff
@@ -0,0 +1,167 @@
+ */
+
+#pragma once
+
+#include <seqan3/alignment/band/static_band.hpp>
+#include <seqan3/alignment/pairwise/detail/alignment_algorithm_cache.hpp>
+#include <seqan3/core/type_traits/basic.hpp>
+#include <seqan3/range/views/slice.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief Manages the alignment and score matrix.
+ * \ingroup alignment_policy
+ *
+ * \tparam alignment_algorithm_t The derived type (seqan3::detail::alignment_algorithm) to be augmented with this
+ *                               CRTP-policy.
+ * \tparam score_matrix_t The type of the alignment score matrix.
+ * \tparam trace_matrix_t The type of the alignment trace matrix.
+ *
+ * \details
+ *
+ * This policy is used to manage the score and trace matrix of the alignment algorithm. On invocation of an alignment
+ * instance the necessary memory is allocated and the corresponding matrix iterators are initialised. These
+ * iterators are used as a global state within this particular alignment instance and are accessed from the alignment
+ * algorithm.
+ *
+ * \remarks The template parameters of this CRTP-policy are selected in the
+ *          seqan3::detail::alignment_configurator::select_matrix_policy when selecting the alignment for the given
+ *          configuration.
+ */
+template <typename alignment_algorithm_t, typename score_matrix_t, typename trace_matrix_t>
+class alignment_matrix_policy
+{
+private:
+    //!\brief Allow alignment algorithm to instantiate this crtp base class.
+    friend alignment_algorithm_t;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alignment_matrix_policy() = default; //!< Defaulted.
+    constexpr alignment_matrix_policy(alignment_matrix_policy const &) = default; //!< Defaulted.
+    constexpr alignment_matrix_policy(alignment_matrix_policy &&) = default; //!< Defaulted.
+    constexpr alignment_matrix_policy & operator=(alignment_matrix_policy const &) = default; //!< Defaulted.
+    constexpr alignment_matrix_policy & operator=(alignment_matrix_policy &&) = default; //!< Defaulted.
+    ~alignment_matrix_policy() = default; //!< Defaulted.
+    //!}
+
+    /*!\brief Allocates the memory of the underlying matrices.
+     * \tparam fst_range_t The type of the first sequence to align; must model std::forward_ranges.
+     * \tparam snd_range_t The type of the second sequence to align; must model std::forward_ranges.
+     * \tparam fst_range_t The type of the first sequence to align.
+     * \tparam sec_range_t The type of the second sequence to align.
+     * \param[in] fst_range The first sequence to align.
+     * \param[in] snd_range The second sequence to align.
+     *
+     * \details
+     *
+     * Initialises the underlying score and trace matrices and sets the respective matrix iterators to the begin of the
+     * corresponding matrix.
+     */
+    template <typename fst_range_t, typename snd_range_t>
+    constexpr void allocate_matrix(fst_range_t && fst_range, snd_range_t && snd_range)
```

Comment:
what the ranges? If someone want's to use a filtered sequence with all Ns removed so be it. Why would I limit that. It should also not be this performance critical since the operations are done on the sequence level and not within the algorithm. Of course some things like getting the size are more expensive because they are now linear instead of constant but I wouldn't limit that in the algorithm